### PR TITLE
feat: use new gateway for pixi search

### PR DESCRIPTION
**Pull request by nichmor**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -326,26 +326,27 @@ tokio = { workspace = true, features = ["rt"] }
...
+# # deno_task_shell = { path = "../deno_task_shell" }
+rattler = { git = "https://github.com/nichmor/rattler", branch = "feat/add-search-for-packages-for-repodata" }
+rattler_conda_types = { git = "https://github.com/nichmor/rattler", branch = "feat/add-search-for-packages-for-repodata" }
+rattler_digest = { git = "https://github.com/nichmor/rattler", branch = "feat/add-search-for-packages-for-repodata" }
+rattler_lock = { git = "https://github.com/nichmor/rattler", branch = "feat/add-search-for-packages-for-repodata" }
+rattler_networking = { git = "https://github.com/nichmor/rattler", branch = "feat/add-search-for-packages-for-repodata" }
+rattler_repodata_gateway = { git = "https://github.com/nichmor/rattler", branch = "feat/add-search-for-packages-for-repodata" }
+rattler_shell = { git = "https://github.com/nichmor/rattler", branch = "feat/add-search-for-packages-for-repodata" }
+rattler_solve = { git = "https://github.com/nichmor/rattler", branch = "feat/add-search-for-packages-for-repodata" }
+rattler_virtual_packages = { git = "https://github.com/nichmor/rattler", branch = "feat/add-search-for-packages-for-repodata" }
```
Comment (baszalmstra) [inline Cargo.toml:338]: This can be removed right?

File: src/cli/search.rs [hunk 2]
```diff
@@ -45,44 +45,50 @@ pub struct Args {
...
+        // sort records by version, get the latest one of each package
+        let records_of_repo: HashMap<String, RepoDataRecord> = repo
+            .into_iter()
+            .sorted_by(|a, b| a.package_record.version.cmp(&b.package_record.version))
+            .rev() // Reverse the iterator so that the latest version comes first
+            .fold(HashMap::new(), |mut acc, record| {
+                acc.entry(record.package_record.name.as_normalized().to_string())
+                    .or_insert(record.clone());
+                acc
+            });
```
Comment (baszalmstra) [inline src/cli/search.rs:89]: Nice. I think you can also do:

```suggestion
		.map(|r| (record.package_record.name.as_normalized().to_string(), record))
		.collect()
```

File: src/cli/search.rs [hunk 3]
```diff
@@ -101,21 +107,31 @@ pub async fn execute(args: Args) -> miette::Result<()> {
...
-    );
+    let config = Config::load_global();
+
+    // Fetch the all names from the repodata using gateway
+    let gateway = config.gateway(auth_client.clone());
+
+    let all_names = gateway
+        .names(channels.clone(), [args.platform, Platform::NoArch])
+        .await
+        .into_diagnostic()?;
```
Comment (baszalmstra) [inline src/cli/search.rs:124]: Should we add a reporter?

File: src/cli/search.rs [hunk 4]
```diff
@@ -280,41 +299,47 @@ fn print_package_info<W: Write>(package: &RepoDataRecord, mut out: W) -> io::Res
...
-    repo_data: Arc<IndexMap<(Channel, Platform), SparseRepoData>>,
+    all_package_names: Vec<String>,
+    repodata_query_func: QF,
     limit: Option<usize>,
     out: W,
-) -> miette::Result<()> {
+) -> miette::Result<()>
+where
+    QF: FnOnce(Vec<MatchSpec>) -> BoxFuture<'static, Result<Vec<RepoData>, GatewayError>>
+        + std::clone::Clone,
```
Comment (baszalmstra) [inline src/cli/search.rs:312]: Instead of boxing you can also make the future generic.

**Review by baszalmstra — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 1 above)
Comment (nichmor) [reply Cargo.toml:338]: not really - we need another rattler release for this

File: src/cli/search.rs (hunk 3 above)
Comment (nichmor) [reply src/cli/search.rs:124]: wrapped it with await_in_progress

File: src/cli/search.rs (hunk 4 above)
Comment (nichmor) [reply src/cli/search.rs]: done!

File: Cargo.toml [hunk 5]
```diff
@@ -326,26 +326,27 @@ tokio = { workspace = true, features = ["rt"] }
...
+# # deno_task_shell = { path = "../deno_task_shell" }
+rattler = { git = "https://github.com/conda/rattler", branch = "main" }
+rattler_conda_types = { git = "https://github.com/conda/rattler", branch = "main" }
+rattler_digest = { git = "https://github.com/conda/rattler", branch = "main" }
+rattler_lock = { git = "https://github.com/conda/rattler", branch = "main" }
+rattler_networking = { git = "https://github.com/conda/rattler", branch = "main" }
+rattler_repodata_gateway = { git = "https://github.com/conda/rattler", branch = "main" }
+rattler_shell = { git = "https://github.com/conda/rattler", branch = "main" }
+rattler_solve = { git = "https://github.com/conda/rattler", branch = "main" }
+rattler_virtual_packages = { git = "https://github.com/conda/rattler", branch = "main" }
```
Comment (baszalmstra) [inline Cargo.toml:338]: This should be removed (or commented out)

**Review by baszalmstra — COMMENTED** [review]
There are a number of linting errors be otherwise LGTM

**Review by baszalmstra — APPROVED** [review]

**Pull request merged**